airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py

```diff
@@ -0,0 +1,321 @@
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        return {"CampaignId": stream_slice["campaign_id"], "ReturnAdditionalFields": self.additional_fields}
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        campaigns = Campaigns(self.client, self.config)
+        for account in Accounts(self.client, self.config).read_records(SyncMode.full_refresh):
+            for campaign in campaigns.read_records(sync_mode=SyncMode.full_refresh, stream_slice={"account_id": account["Id"]}):
+                yield {"campaign_id": campaign["Id"], "account_id": account["Id"]}
+
+        yield from []
+
+
+class Ads(BingAdsStream):
+    """
+    Retrieves the ads for all provided accounts.
+    API doc: https://docs.microsoft.com/en-us/advertising/campaign-management-service/getadsbyadgroupid?view=bingads-13
+    Ad schema: https://docs.microsoft.com/en-us/advertising/campaign-management-service/ad?view=bingads-13
+    """
+
+    data_field: str = "Ad"
+    service_name: str = "CampaignManagement"
+    operation_name: str = "GetAdsByAdGroupId"
+    additional_fields: str = "ImpressionTrackingUrls Videos LongHeadlines"
+    ad_types: Iterable[str] = [
+        "Text",
+        "Image",
+        "Product",
+        "AppInstall",
+        "ExpandedText",
+        "DynamicSearch",
+        "ResponsiveAd",
+        "ResponsiveSearch",
+    ]
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        return {
+            "AdGroupId": stream_slice["ad_group_id"],
+            "AdTypes": {"AdType": self.ad_types},
+            "ReturnAdditionalFields": self.additional_fields,
+        }
+
+    def stream_slices(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        ad_groups = AdGroups(self.client, self.config)
+        for slice in ad_groups.stream_slices(sync_mode=SyncMode.full_refresh):
+            for ad_group in ad_groups.read_records(sync_mode=SyncMode.full_refresh, stream_slice=slice):
+                yield {"ad_group_id": ad_group["Id"], "account_id": slice["account_id"]}
+        yield from []
+
+
+class SourceBingAds(AbstractSource):
+    """
+    Source implementation of Bing Ads API. Fetches advertising data from accounts
+    """
+
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Any]:
+        try:
+            client = Client(**config)
+            Accounts(client, config).read_records(SyncMode.full_refresh)
+        except Exception as error:
+            return False, error
```

Comment:
read_records is an iterator and its call will be executed in some loop only.
I guess to you can create a instance of the class Client only. Its constructor method tries to make a connection with the API.
Or you can use a loop:
```
for record in Accounts(client, config).read_records(SyncMode.full_refresh):
   break
```

Reply:
Can we check available account_ids here? and make a simple validation of their values.

Reply:
done